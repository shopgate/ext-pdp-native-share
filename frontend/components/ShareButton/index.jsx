import React, { Component } from 'react';
import ShareIcon from '@shopgate/pwa-ui-ios/icons/ShareIcon';
import MaterialShareIcon from '@shopgate/pwa-ui-material/icons/MaterialShareIcon';
import Ripple from '@shopgate/pwa-ui-shared/Ripple';
import PropTypes from 'prop-types';
import styles from './style';
import getConfig from '../../helpers/getConfig';
import connect from '../../connector';

const { androidShareSVG } = getConfig();
const { iOSShareSVG } = getConfig();
/**
 * ShareButton component
 */
class ShareButton extends Component {
  static propTypes = {
    iOSTheme: PropTypes.func.isRequired,
    shareItem: PropTypes.func.isRequired,
    shareParams: PropTypes.shape(),
  };

  static defaultProps = {
    shareParams: null,
  };
  /**
   * Handles the share button click
   * Show's share screen for app
   * @param {Object} event The click event object
   */
  handleClick = () => {
    this.props.shareItem();
  }
  /**
   * Renders the share icon depending on theme
   * @returns {JSX}
   */
  renderIcon() {
    if (this.props.iOSTheme()) {
      return <ShareIcon svg={iOSShareSVG} />;
    }
    return <MaterialShareIcon svg={androidShareSVG} />;
  }
  /**
   * Renders the components
   * @returns {JSX}
   */
  render() {
    if (!this.props.shareParams) {
      return null;
    }
    const buttons = this.props.iOSTheme() ? styles.iOSButtons : styles.androidButtons;
    const className = this.props.iOSTheme() ? styles.buttonFlat : styles.button;
    return (
      <div className={`${buttons}`}>
        <button
          className={`${className}`}
          onClick={this.handleClick}
          data-test-id="shareIcon"
        >
          <Ripple className={`${styles.ripple}`}>
            {this.renderIcon()}
          </Ripple>
        </button>
      </div>
    );
  }
}

export default connect(ShareButton);
