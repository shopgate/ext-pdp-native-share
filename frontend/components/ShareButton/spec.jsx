import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';

const mockedConfig = {
  gmdIcon: 'gmd',
  iOSIcon: 'ios',
};

jest.mock('../../helpers/getConfig', () => () => mockedConfig);
const mockedParams = {
  mockedShareParams: {
    title: 'title',
    imageURL: 'imageURL',
    deepLink: 'deepLink',
  },
};

let mockedIsIOS = true;
jest.mock('@shopgate/pwa-extension-kit', () => ({
  env: {
    isIOSTheme: () => mockedIsIOS
  }
}));

jest.mock('../../selectors/index', () => ({
  getShareParams: () => mockedParams.mockedShareParams,
}));

jest.mock('@shopgate/pwa-extension-kit/connectors', () => ({
  withPageProductId: WrappedComponent => () => <WrappedComponent productId="foo" />,
}));

/**
 * Mocked function for shareItem
 * @param {Object} mockedShareParams mocked parameters
 */
const mockedShareItem = jest.fn();
jest.mock('@shopgate/pwa-core/commands/shareItem', () => () => mockedShareItem());

describe('ShareButton', () => {
  // eslint-disable-next-line require-jsdoc
  const makeComponent = () => {
    // eslint-disable-next-line global-require
    const ShareButton = require('./index').default;
    return mount((
      <Provider store={configureStore()({})}>
        <ShareButton />
      </Provider>
    ));
  };

  it('should render when shareParams exist and handle click', () => {
    const component = makeComponent();
    expect(component.find('ShareButton').props().shareParams).toMatchObject(mockedParams.mockedShareParams);
    expect(component).toMatchSnapshot();

    component.find('ShareButton').instance().handleClick();
    expect(mockedShareItem).toHaveBeenCalled();
  });

  it('should render with gmd icon relying on theme', () => {
    mockedIsIOS = false;
    const component = makeComponent();
    expect(component.find('ShareButton').props().shareParams).toMatchObject(mockedParams.mockedShareParams);
    expect(component).toMatchSnapshot();
  });

  it('should render with ios icon relying on theme', () => {
    mockedIsIOS = true;
    const component = makeComponent();
    expect(component.find('ShareButton').props().shareParams).toMatchObject(mockedParams.mockedShareParams);
    expect(component).toMatchSnapshot();
  });

  it('should render with ios relying on config', () => {
    mockedIsIOS = false;
    mockedConfig.gmdIcon = 'ios';
    const component = makeComponent();
    expect(component.find('ShareButton').props().shareParams).toMatchObject(mockedParams.mockedShareParams);
    expect(component).toMatchSnapshot();
  });

  it('should render with gmd relying on config', () => {
    mockedIsIOS = true;
    mockedConfig.iOSIcon = 'gmd';
    const component = makeComponent();
    expect(component.find('ShareButton').props().shareParams).toMatchObject(mockedParams.mockedShareParams);
    expect(component).toMatchSnapshot();
  });

  it('should not render when deeplink is undefined', () => {
    mockedParams.mockedShareParams.deepLink = undefined;
    const component = makeComponent();
    expect(component.html()).toBe(null);
  });
});
