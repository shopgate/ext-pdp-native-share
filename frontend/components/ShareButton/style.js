import { css } from 'glamor';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';

const androidButtons = css({
  position: 'absolute',
  right: 130,
  padding: 8,
  top: -30,
  display: 'flex',
  justifyContent: 'flex-end',
  alignItems: 'center',
  zIndex: 1,
}).toString();

const iOSButtons = css({
  position: 'absolute',
  right: 68,
  top: -30,
  display: 'flex',
  justifyContent: 'flex-end',
  alignItems: 'center',
  zIndex: 1,
}).toString();

const buttonProto = {
  display: 'block',
  position: 'relative',
  background: '#fff',
  borderRadius: '50%',
  padding: 0,
  fontSize: 20,
  lineHeight: 1,
  color: themeConfig.colors.accent,
  outline: 0,
};

const buttonFlat = css(buttonProto).toString();

const button = css({
  ...buttonProto,
  boxShadow: '0 8px 13px rgba(0, 0, 0, 0.25)',
}).toString();

const ripple = css({
  padding: 10,
}).toString();

export default {
  androidButtons,
  iOSButtons,
  buttonFlat,
  button,
  ripple,
};

