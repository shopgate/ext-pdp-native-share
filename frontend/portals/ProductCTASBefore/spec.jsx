import React from 'react';
import { mount } from 'enzyme';
import ShareButton from '../../components/ShareButton';

// eslint-disable-next-line require-jsdoc
const MockedShareButton = () => (<div>ShareButton</div>);
jest.mock('../../components/ShareButton', () => MockedShareButton);
let mockedIsIOS = true;
jest.mock('@shopgate/pwa-extension-kit', () => ({
  env: {
    isIOSTheme: () => mockedIsIOS
  }
}));

describe('IosShareButton', () => {
  // eslint-disable-next-line global-require
  const ShareButton = require('./index').default;
  it('should render ShareButton', () => {
    const component = mount(<ShareButton />);
    expect(component.find(MockedShareButton).exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });

  it('should render null', () => {
    mockedIsIOS = false;
    const component = mount(<ShareButton />);
    expect(component.html()).toBe(null);
  });
});
