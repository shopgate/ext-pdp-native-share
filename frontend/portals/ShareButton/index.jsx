import React from 'react';
import ShareButton from '../../components/ShareButton';

export default props => <ShareButton {...props} />;
