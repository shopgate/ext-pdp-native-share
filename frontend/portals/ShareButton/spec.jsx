import React from 'react';
import { mount } from 'enzyme';
import ShareButton from './index';

jest.mock('../../components/ShareButton', () => () => <div>ShareButton</div>);

describe('ShareButton', () => {
  it('should render ShareButton', () => {
    const component = mount(<ShareButton iOSTheme={false} />);
    expect(component).toMatchSnapshot();
  });
});
