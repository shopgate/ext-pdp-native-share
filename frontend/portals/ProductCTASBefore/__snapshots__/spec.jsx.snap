// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IosShareButton should render ShareButton 1`] = `
<_default>
  <div
    className="css-1npkug1"
  >
    <MockedShareButton>
      <div>
        ShareButton
      </div>
    </MockedShareButton>
  </div>
</_default>
`;
